File: src/gitph/git/graph.py
from collections import defaultdict

from gitph.domain.models import CommitSummary, GitRef, GraphEdge, GraphModel, GraphRow


class GitGraphBuilder:
    """Builds a render-neutral commit lane graph from topo-ordered commits."""

    def build(
        self,
        commits: tuple[CommitSummary, ...],
        refs: tuple[GitRef, ...],
    ) -> GraphModel:
        refs_by_oid: dict[str, list[GitRef]] = defaultdict(list)
        for ref in refs:
            refs_by_oid[ref.display_oid].append(ref)

        active_lanes: list[str | None] = []
        rows: list[GraphRow] = []
        max_lanes = 0

        for row_index, commit in enumerate(commits):
            lane = self._claim_lane(active_lanes, commit.oid)
            parents = commit.parents
            if parents:
                active_lanes[lane] = parents[0]
            else:
                active_lanes[lane] = None

            edges: list[GraphEdge] = []
            for parent_index, parent_oid in enumerate(parents):
                parent_lane = self._ensure_lane(active_lanes, parent_oid, preferred_lane=lane)
                edges.append(
                    GraphEdge(
                        from_oid=commit.oid,
                        to_oid=parent_oid,
                        from_lane=lane,
                        to_lane=parent_lane,
                        color_index=parent_lane % 8,
                        kind="merge" if parent_index > 0 else "parent",
                    )
                )

            self._trim_lanes(active_lanes)
            max_lanes = max(max_lanes, len(active_lanes), lane + 1)
            rows.append(
                GraphRow(
                    commit=commit,
                    row_index=row_index,
                    lane=lane,
                    refs=tuple(refs_by_oid.get(commit.oid, ())),
                    edges=tuple(edges),
                    active_lanes=tuple(active_lanes),
                )
            )

        return GraphModel(rows=tuple(rows), max_lanes=max_lanes)

    @staticmethod
    def _claim_lane(active_lanes: list[str | None], oid: str) -> int:
        if oid in active_lanes:
            return active_lanes.index(oid)
        for index, value in enumerate(active_lanes):
            if value is None:
                active_lanes[index] = oid
                return index
        active_lanes.append(oid)
        return len(active_lanes) - 1

    @staticmethod
    def _ensure_lane(active_lanes: list[str | None], oid: str, preferred_lane: int) -> int:
        if oid in active_lanes:
            return active_lanes.index(oid)
        if preferred_lane < len(active_lanes) and active_lanes[preferred_lane] == oid:
            return preferred_lane
        for index, value in enumerate(active_lanes):
            if value is None:
                active_lanes[index] = oid
                return index
        active_lanes.append(oid)
        return len(active_lanes) - 1

    @staticmethod
    def _trim_lanes(active_lanes: list[str | None]) -> None:
        while active_lanes and active_lanes[-1] is None:
            active_lanes.pop()

File: src/main/git/graph.ts
import type { CommitSummary, GitRef, GraphEdge, GraphModel, GraphRow } from '../../shared/contracts'

/**
 * A lane slot holds the OID of the commit expected further down that column.
 *
 * The primary slot is the column where that commit will render its node; a
 * non-primary slot is a reserved tail: a converging branch keeps its column
 * occupied until the parent's row so the renderer can run the branch line down
 * it without another branch reusing the column mid-flight.
 */
interface LaneSlot {
  oid: string
  primary: boolean
}

/**
 * @brief Produces a render-neutral lane layout for topologically ordered commits.
 *
 * Responsibility: assign stable lanes and parent connections once in the main
 * process so every renderer can remain a lightweight view of the same graph.
 */
export class GitGraphBuilder {
  build(commits: readonly CommitSummary[], refs: readonly GitRef[]): GraphModel {
    const refsByOid = new Map<string, GitRef[]>()
    for (const ref of refs) {
      const bucket = refsByOid.get(ref.displayOid) ?? []
      bucket.push(ref)
      refsByOid.set(ref.displayOid, bucket)
    }

    const lanes: Array<LaneSlot | null> = []
    const rows: GraphRow[] = []
    let maxLanes = 0

    commits.forEach((commit, rowIndex) => {
      const lane = this.takeLane(lanes, commit.oid)
      const firstParentOid = commit.parents[0] ?? null
      const firstParentLane = firstParentOid ? this.findPrimaryLane(lanes, firstParentOid) : -1

      if (firstParentOid === null) {
        lanes[lane] = null
      } else if (firstParentLane >= 0) {
        // Converging first parent: this branch ends at that parent, so hold the
        // column as a reserved tail until the parent's row is emitted.
        lanes[lane] = { oid: firstParentOid, primary: false }
      } else {
        lanes[lane] = { oid: firstParentOid, primary: true }
      }

      const edges: GraphEdge[] = commit.parents.map((parentOid, parentIndex) => {
        const parentLane = this.ensurePrimaryLane(lanes, parentOid)
        return {
          fromOid: commit.oid,
          toOid: parentOid,
          fromLane: lane,
          toLane: parentLane,
          colorIndex: parentLane % 8,
          kind: parentIndex > 0 ? 'merge' : 'parent'
        }
      })

      while (lanes.length > 0 && lanes.at(-1) === null) lanes.pop()
      maxLanes = Math.max(maxLanes, lanes.length, lane + 1)
      rows.push({
        commit,
        rowIndex,
        lane,
        refs: [...(refsByOid.get(commit.oid) ?? [])],
        edges,
        activeLanes: lanes.map((slot) => slot?.oid ?? null)
      })
    })

    return { rows, maxLanes }
  }

  /**
   * Returns the lane where this commit renders and releases every slot that
   * was waiting for it (its primary column plus any converging tails).
   */
  private takeLane(lanes: Array<LaneSlot | null>, oid: string): number {
    let claimed = -1
    for (let index = 0; index < lanes.length; index += 1) {
      const slot = lanes[index]
      if (slot?.oid !== oid) continue
      if (claimed < 0 || slot.primary) claimed = index
      lanes[index] = null
    }
    if (claimed >= 0) return claimed
    const free = lanes.indexOf(null)
    if (free >= 0) return free
    lanes.push(null)
    return lanes.length - 1
  }

  private findPrimaryLane(lanes: Array<LaneSlot | null>, oid: string): number {
    return lanes.findIndex((slot) => slot?.oid === oid && slot.primary)
  }

  private ensurePrimaryLane(lanes: Array<LaneSlot | null>, oid: string): number {
    const existing = this.findPrimaryLane(lanes, oid)
    if (existing >= 0) return existing
    const slot: LaneSlot = { oid, primary: true }
    const free = lanes.indexOf(null)
    if (free >= 0) {
      lanes[free] = slot
      return free
    }
    lanes.push(slot)
    return lanes.length - 1
  }
}
